From b36c2d928f26780ff4137716fb470ea1999050d4 Mon Sep 17 00:00:00 2001
From: Portisch <hugo.portisch@yahoo.de>
Date: Thu, 4 Apr 2019 07:16:58 +0200
Subject: [PATCH] ScreenshotAML: update amvideocap frame capture to
 CAP_FLAG_AT_END method This patch will ensure a sync frame capture from the
 video frame queue

---
 xbmc/utils/ScreenshotAML.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/xbmc/utils/ScreenshotAML.cpp b/xbmc/utils/ScreenshotAML.cpp
index 8b67673..9e221b5 100644
--- a/xbmc/utils/ScreenshotAML.cpp
+++ b/xbmc/utils/ScreenshotAML.cpp
@@ -19,8 +19,11 @@
 #define AMVIDEOCAP_IOC_MAGIC  'V'
 #define AMVIDEOCAP_IOW_SET_WANTFRAME_WIDTH      _IOW(AMVIDEOCAP_IOC_MAGIC, 0x02, int)
 #define AMVIDEOCAP_IOW_SET_WANTFRAME_HEIGHT     _IOW(AMVIDEOCAP_IOC_MAGIC, 0x03, int)
+#define AMVIDEOCAP_IOW_SET_WANTFRAME_AT_FLAGS   _IOW(AMVIDEOCAP_IOC_MAGIC, 0x06, int)
 #define AMVIDEOCAP_IOW_SET_CANCEL_CAPTURE       _IOW(AMVIDEOCAP_IOC_MAGIC, 0x33, int)
 
+#define CAP_FLAG_AT_END 2
+
 // capture format already defaults to GE2D_FORMAT_S24_RGB - no need to pull in all the ge2d headers :)
 
 #define CAPTURE_DEVICEPATH "/dev/amvideocap0"
@@ -42,6 +45,7 @@ void CScreenshotAML::CaptureVideoFrame(unsigned char *buffer, int iWidth, int iH
       // configure destination
       ioctl(captureFd, AMVIDEOCAP_IOW_SET_WANTFRAME_WIDTH, stride / 3);
       ioctl(captureFd, AMVIDEOCAP_IOW_SET_WANTFRAME_HEIGHT, iHeight);
+      ioctl(captureFd, AMVIDEOCAP_IOW_SET_WANTFRAME_AT_FLAGS, CAP_FLAG_AT_END);
       readSize = pread(captureFd, videoBuffer, buffSize, 0);
     }
 
-- 
1.9.1

